File: DemoServer/AdditionalLanguages/go/RelatedDocuments/createRelatedDocuments/CreateRelatedDocuments.go
package main

//region Usings
import "github.com/ravendb/ravendb-go-client"
//endregion

var globalDocumentStore *ravendb.DocumentStore

func main() {
    createDocumentStore()
    createDatabase()
    createRelatedDocuments("someProductName","someSupplierName","somePhoneNumber")
    globalDocumentStore.Close()
}

func createDocumentStore() (*ravendb.DocumentStore, error) {
    if globalDocumentStore != nil {
        return globalDocumentStore, nil
    }
    urls := []string{"http://localhost:8080"}
    store := ravendb.NewDocumentStore(urls, "testGO")
    err := store.Initialize()
    if err != nil {
        return nil, err
    }
    globalDocumentStore = store
    return globalDocumentStore, nil
}

func createDatabase() {
    databaseRecord := ravendb.NewDatabaseRecord()
    databaseRecord.DatabaseName = "testGO"
    createDatabaseOperation := ravendb.NewCreateDatabaseOperation(databaseRecord, 1)
    var err = globalDocumentStore.Maintenance().Server().Send(createDatabaseOperation)
    if err != nil {
        fmt.Printf("d.store.Maintenance().Server().Send(createDatabaseOperation) failed with %s\n", err)
    }
}

type Company struct {
    ID      string
    Name    string
    Phone   string
    Contact *Contact
}

type Contact struct {
    Name  string
    Title string
}

//region Demo
func createRelatedDocuments(productName, supplierName, supplierPhone string) error {

    //region Step_1
    supplier := &Supplier {
        Name:  supplierName,
            Phone: supplierPhone,
    }    
    category := &Category {
        Name:        "NoSQL Databases",
        Description: "Non-relational databases",
    }
    //endregion    
    //region Step_2
    product := &Product{
        Name: productName,
    }    
    //endregion
    
    session, err := globalDocumentStore.OpenSession("")
    if err != nil {
        return err
    }
    defer session.Close()

    //region Step_3
    err = session.Store(supplier)
    if err != nil {
        return err
    }    
    err = session.Store(category)
    if err != nil {
        return err
    }
    //endregion

    //region Step_4
    product.Supplier = supplier.ID
    product.Category = category.ID
    //endregion

    //region Step_5
    err = session.Store(product)
    if err != nil {
        return err
    }
    //endregion
     
    //region Step_6
    err = session.SaveChanges()
    if err != nil {
        return err
    }
    //endregion
  
    return nil 
}
//endregion

File: DemoServer/AdditionalLanguages/python/relatedDocuments/createRelatedDocuments/createRelatedDocuments.py
from demo_example import RunParamsBase, Example
from models import Supplier, Category, Product

class RunParams(RunParamsBase):
    def __init__(self, supplier_name: str, supplier_phone: str, product_name: str):
        self.supplier_name = supplier_name
        self.supplier_phone = supplier_phone
        self.product_name = product_name


class CreateRelatedDocuments(Example):
    def run(self, run_params: RunParams) -> Product:
        supplier_name = run_params.supplier_name
        supplier_phone = run_params.supplier_phone
        product_name = run_params.product_name

        #region Demo
        #region Step_1
        supplier = Supplier()
        supplier.Name = supplier_name
        supplier.Phone = supplier_phone

        category = Category()
        category.Name = "NoSQL Databases"
        category.Description = "Non-relational databases"
        #endregion

        #region Step_2
        product = Product(Name=product_name)
        #endregion

        with self.document_store_holder.store().open_session() as session:
            #region Step_3
            session.store(supplier)
            session.store(category)
            #endregion

            #region Step_4
            product.Supplier = supplier.Id
            product.Category = category.Id
            #endregion

            #region Step_5
            session.store(product, "products/34-A")
            #endregion

            #region Step_6
            session.save_changes()
            #endregion

        #endregion
        return product
